rpmlb/builder/koji.py

```diff
@@ -0,0 +1,173 @@
+"""Builder interface for the Koji build service."""
+
+import re
+from pathlib import Path
+from typing import Iterator, Optional, Sequence
+
+from ..utils import run_cmd, run_cmd_with_capture
+from ..work import Work
+from .base import BaseBuilder
+
+
+class KojiBuilder(BaseBuilder):
+    """Builder interface for the Koji build service."""
+
+    DEFAULT_TARGET_FORMAT = 'sclo{epel}-{collection}-rh-el{epel}'
+
+    def __init__(self):
+        """Initialize the slots"""
+
+        self.collection = None  #: The collection to build
+        self.epel = None  #: Version of EPEL to build for
+        self.profile = None  #: The koji configuration profile to use
+        #: The format string for construction of the actual template name
+        self.target_format = type(self).DEFAULT_TARGET_FORMAT
+
+    @property
+    def base_command(self) -> Sequence[str]:
```

Comment:
> why you are being so explicit about it instead of clearly explaining your intentions

I am sorry. I was rude.

The class inheritance `is-a` is related to the template design pattern.

cli.py Sample.

```
# I intended factory method pattern in this line.
# People do not want to know how the builder instance was created,
# Also do not want to know which type of builder class and the detail of the instance.
#
# Factory method
# https://en.wikipedia.org/wiki/Factory_method_pattern#Python
#
# On only this line, people recognize giving build type as string.
# And after that, do not want to recognize sub class of the builder class.
builder = BaseBuilder.get_instance(A_BUILD_TYPE)

# People only want to know the "Builder" class (BaseBuilder)'s behavior (methods, properties)
builder.run()

# People can use the methods that are defined on BaseBuilder.py
# But can not access, use the methods/properties that are only defined on specified sub class of Builder. Using below methods is ok.
builder.before()
builder.after()
builder.prepare()

# But below case is not ok. Technically may be able to access in Python.
builder.a_sub_class_specified_method()

```

Another example.

```
builders = [
    BaseBuilder.get_instance(BUILD_TYPE_A)
    BaseBuilder.get_instance(BUILD_TYPE_B)
    BaseBuilder.get_instance(BUILD_TYPE_B)
]
  
for builder in builders:
    builder.run()
    somethings.
```

Another example.

```
builders = [
    BaseBuilder.get_instance(BUILD_TYPE_A)
    BaseBuilder.get_instance(BUILD_TYPE_B)
    BaseBuilder.get_instance(BUILD_TYPE_B)
]
  
for builder in builders:
    builder.run()
    somethings.
```

So, in this pattern, "a public interface of a subclass" (that can be `has-a` methods/properties) can not be accessed from outside of the class.

another example of Java

```
List list_a = new ArrayList()
List list_b = new Vector()

# as size() is defined in List interface, we can access.
list_a.size()
list_b.size()
```

> "one for any possible subclasses of this given builder"
The one (method) is called "protected" method, as you know.
The extend public interface(methods/properties) are only used from inherited class.

`_method_name`: protected method
`__method_name`: private method
https://stackoverflow.com/questions/11483366/protected-method-in-python

You do not like underscore method to describe the it is protected method?

”Polymorphism”
This same method name and different content is called Polymorphism.
https://en.wikipedia.org/wiki/Polymorphism_(computer_science)

"Loose coupling" is also a key idea.
We do not want to know many class from a class.
https://en.wikipedia.org/wiki/Loose_coupling

"Encapsulation" is also.
We only know only necessary things from outside.
https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)